File: java/allergies/src/main/java/Allergies.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Allergies {
    private final List<Allergen> allergies;

    public Allergies(int score) {
        List<Allergen> allergies = new ArrayList<>();
        for (Allergen allergen : Allergen.values()) {
            if ((score & allergen.getScore()) != 0) {
                allergies.add(allergen);
            }
        }

        this.allergies = Collections.unmodifiableList(allergies);
    }

    public Boolean isAllergicTo(Allergen allergen) {
        return allergies.contains(allergen);
    }

    public List<Allergen> getList() {
        return allergies;
    }
}

File: cpp/allergies/allergies.cpp
#include "allergies.h"

#include <algorithm>
#include <array>
#include <iterator>

namespace allergies {

namespace {

const std::array<std::string_view, 8> allergens{
    "eggs",
    "peanuts",
    "shellfish",
    "strawberries",
    "tomatoes",
    "chocolate",
    "pollen",
    "cats",
};

} // anonymous namespace

allergy_test::allergy_test(std::size_t score)
    : score_{score}
{
}

auto allergy_test::is_allergic_to(std::string_view allergen) const -> bool
{
    return is_allergic_to(std::find(allergens.cbegin(), allergens.cend(), allergen) - allergens.cbegin());
}

auto allergy_test::get_allergies() const -> std::unordered_set<std::string>
{
    std::unordered_set<std::string> allergies;

    for (std::size_t allergen_score = 0; allergen_score < allergens.size(); ++allergen_score) {
        if (is_allergic_to(allergen_score)) {
            allergies.emplace(allergens[allergen_score]);
        }
    }

    return allergies;
}

}  // namespace allergies
